### ISSUE TITLE
OOM during typecheking in 6.1.13

### ISSUE THREAD

**Issue by PaulBlanche** [styled-components/styled-components#4350]
### System:
- OS: Linux 6.10 Arch Linux
- CPU: (8) x64 11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
- Memory: 8.54 GB / 15.35 GB
- Container: Yes
- Shell: 5.9 - /bin/zsh

### Binaries:
- Node: 22.5.1 - /usr/bin/node
- Yarn: 1.22.22 - /usr/bin/yarn
- npm: 10.8.2 - /usr/bin/npm

### npmPackages:
- babel-plugin-styled-components: 2.1.4 => 2.1.4
- styled-components: ^6.1.13 => 6.1.13

## Steps to reproduce

We have a large codebase using styled-component (~5000 `styled` function usage). The repo is private so i can't share it. We typically use styled components like this : 

```tsx
import * as React from 'react'
import styled from 'styled-components'

export type InputProps = Omit<React.InputHTMLAttributes<HTMLInputElement>, 'type'> & {
  type?: 'text' | 'search' | 'email' | 'password'
  status?:  'error' | 'success' | 'default'
  small?: boolean
}

const InputComponent = React.forwardRef<HTMLInputElement, InputProps>(function InputComponent(
  { status, small, className, ...inputProps },
  inputRef,
) {
  return (
    <Wrapper className={className} small={small}>
      <Input $disabled={inputProps.disabled} status={status}>
        <TextInput ref={inputRef} type="text" {...inputProps} />
      </Input>
    </Wrapper>
  )
})

export const Input = styled(InputComponent)``

const Wrapper = styled.div<{ small?:boolean }>`
  ...
`

const BaseInputWrapper = styled.div`
  ...
`

type InputProps = {
    $disabled?: boolean | undefined;
    status?: Status | undefined;
}

const InputWrapper = styled(BaseInput)<InputProps>`
  ...
`

const TextInput = styled.input`
  ...
`

```

When typechecking with `tsc` we get an OOM. When checking hotspots with ` @typescript/analyze-trace` they are all over the place on styled components.

I dug in the types of styled-component, and the OOM stopped when i replaced the `Substitute` in the [interface `Styled`](https://github.com/styled-components/styled-components/blob/6f6db180bd0ec89bd4342dc4b8f1eae0b34d8dca/packages/styled-components/src/constructors/constructWithOptions.ts#L44C1-L82C2) with a simple `&` (less typesafe, but overall same "meaning").

To merge the props with `Substitute` given the props we use (since we rely on `React.InputHTMLAttributes<HTMLInputElement>`), typescript has to do a lot of work for each instance which might be the cause of the OOM. 

So the types that infer the props of a styled components seems to be to heavy for our codebase. For now we resorted to using the v6 implementation and v5 types to avoid the OOM.

Is our usage of styled-component too exotic ? Do we have to change our way of using it ? Or is styled-component supposed to support this kind of usage for a large codebase ?

**Comment by joesaunderson** [comment]
+1 to this issue, we have also recently hit the same thing.

**Comment by Codexphere92** [comment]
hey @PaulBlanche, were you able to resolve this issue due to `Substitute`?

**Comment by PaulBlanche** [comment]
Not really, the only way i found to make it work with our codebase was to create a facade module that exports the implementation of v6 with the types of v5 to avoid the OOM.

**Comment by Codexphere92** [comment]
it seems like the v6 types are assuming every element or component wrapped with styled() can potentially be a polymorphic component and there are some heavy types defined for polymorphic components. maybe the types can be improved by avoiding some type checks unless forwardedAs or as is used. @quantizor is there room to improve the types even though the project is in maintenance mode?

**Comment by quantizor** [comment]
@Codexphere92 Good observation—you're right that every `styled()` call currently resolves the full polymorphic type machinery regardless of whether `as`/`forwardedAs` is used.

## What changed in 6.4

- `Substitute` got a short-circuit (`keyof B extends never ? A : FastOmit<A, keyof B> & B`) that skips `FastOmit` for simple components with no extra props.
- `OverrideStyle` was added for React 19 compat and accounts for ~22% of type instantiations (measured below). It can't be simplified without breaking `exactOptionalPropertyTypes` or JSX overload resolution.

## Experimental scaling data (6.4, TS 5.7, Apple M-series)

We ran controlled scaling experiments with realistic styled components (custom props, interpolations, 5% `as` usage, 3-deep inheritance chains):

| Scenario | Components | Instantiations | Memory | Check time |
|----------|-----------|----------------|--------|------------|
| Library baseline (0 usages) | 0 | 887K | 490MB | 1.7s |
| Realistic x1000 | 1000 | 1.00M | 548MB | 2.3s |
| Realistic x3000 | 3000 | 1.08M | 599MB | 4.0s |
| Diverse x5000 | 5000 | 952K | 545MB | 6.8s |
| **no `skipLibCheck` x1000** | 1000 | **1.49M** | **871MB** | 4.7s |
| **no `skipLibCheck` x3000** | 3000 | **1.57M** | **928MB** | 6.5s |

**Key finding: `skipLibCheck: false` nearly doubles instantiations and memory.** This is likely the primary trigger for OOM in large codebases.

## Ablation: what costs the most

| Experiment | Instantiations (3000 realistic) | Delta |
|------------|--------------------------------|-------|
| Full types (baseline) | 1.08M | — |
| `OverrideStyle` → identity | 845K | **-22%** |
| Polymorphic overload removed | 1.01M | **-6%** |

`OverrideStyle` is the biggest single type cost center. The polymorphic generic overload itself adds only ~6%—TS caches the defaulted generic well when `as` isn't provided.

## TypeScript team position

We investigated filing a TS feature request for lazy/deferred overload resolution. The relevant existing issues are:

- [TS #43717](https://github.com/microsoft/TypeScript/issues/43717) (polymorphic JSX perf)—**closed as "Not a Defect"**. Ryan Cavanaugh benchmarked 0.48ms/element in TS 5.4 and deemed it acceptable.
- [TS #34801](https://github.com/microsoft/TypeScript/issues/34801) (MUI slow)—closed as "External".
- [TS #41263](https://github.com/microsoft/TypeScript/issues/41263) (cache utility types)—closed as "Design Limitation".

The TS team considers the per-element cost of polymorphic components to be within acceptable bounds. A new issue would be closed.

## TypeScript native preview (tsgo) benchmarks

We tested [`@typescript/native-preview`](https://www.npmjs.com/package/@typescript/native-preview) (7.0.0-dev.20260323.1) against the same scaling tests. **Error output is identical to tsc**—same single diagnostic on the same line.

| Scenario | tsc Check | tsgo Check | **Speedup** | tsc Memory | tsgo Memory | **Mem Savings** |
|----------|-----------|------------|-------------|------------|-------------|-----------------|
| Realistic 1000 | 2.23s | 0.80s | **2.8x** | 539MB | 315MB | **-42%** |
| Realistic 3000 | 3.86s | 1.55s | **2.5x** | 542MB | 324MB | **-40%** |
| Diverse 5000 | 7.10s | 3.06s | **2.3x** | 548MB | 328MB | **-40%** |
| no skipLibCheck 3000 | 6.54s | 1.69s | **3.9x** | 889MB | 438MB | **-51%** |

The worst case (3000 components, `skipLibCheck: false`) drops from 6.5s/889MB to **1.7s/438MB**. At these numbers, even 10,000+ styled components should stay well within memory limits.

## Practical mitigations (today, in order of least to most effort)

1. **Enable `skipLibCheck: true`** (the single biggest win—halves memory usage)
2. **Try `@typescript/native-preview`** (`npm i -D @typescript/native-preview && npx tsgo`)—2.5–4x faster with 40–50% less memory, error-compatible with tsc
3. **Increase Node heap**: `NODE_OPTIONS=--max-old-space-size=8192` in your tsconfig or IDE settings
4. **Split large projects** into TS project references so each project checks fewer files
The 6.4 prerelease is available via `npm i styled-components@test` ([Version Packages PR #5687](https://github.com/styled-components/styled-components/pull/5687)) if you'd like to benchmark against your codebase.

**Comment by Stanzilla** [comment]
Adding real-world data from an in-progress 6.4 migration on a large monorepo, in case it's useful for triage and for others arriving here from search. We're not done yet, but we're making steady progress and most of the wall-time wins so far have come from the consumer side.

## Setup

- Nx + pnpm monorepo, ~220 projects, Next.js app + many shared packages
- styled-components: 5.3.11 → 6.4.0 (work in progress on a long-lived branch)
- TypeScript: tsc 5.x and `@typescript/native-preview` (tsgo) 7.0.0-dev.20260425.1, identical findings on both
- `skipLibCheck: true` everywhere, `--noEmit`
- ~thousands of `styled.*` call sites across the workspace

## Workspace-level numbers (web-next, single tsgo run)

After the 6.4 upgrade, before any consumer-side mitigations:

| Metric         | v5.3.11 | v6.4.0  | Delta |
| -------------- | ------- | ------- | ----- |
| Types          | ~3.0M   | 6.6M    | ~2.2× |
| Instantiations | ~22M    | 54.7M   | ~2.5× |
| Symbols        | similar | similar | flat  |

Consistent with the "v6 base type tax" reading, every `styled` call pays more than in v5, and the multiplier compounds at scale. The 6.4 `Substitute` short-circuit definitely helped vs 6.1 (we tried both), but didn't bring us back to v5 levels.

## What's been moving the needle for us — all consumer-side, no library changes

### 1. `theme: any` on theme helpers (biggest single win)

We had a set of helpers shaped like `Record<Keys, ThemedProp>` where `ThemedProp = (props: { theme: ITheme }) => string`. When `theme` was typed strictly as `ITheme`, every styled interpolation that used a helper paid a `DefaultTheme extends ITheme` resolution against our augmented `DefaultTheme`. Loosening just the helpers' `theme` parameter back to `any` (the augmentation stays) recovered seconds of check time on hotspot files.

Lesson: with v6, every `ExecutionContext`-typed function inside a `styled` interpolation re-resolves `DefaultTheme`. Strict typing inside helper modules amplifies that linearly.

### 2. Widen large literal-keyed exports that are imported broadly — with a real downside

We had a generated `Record<LiteralKey, MessageDescriptor>` with ~25k keys (`react-intl` `sharedMessages.ts`) that turned out to be the #1 single-file hotspot post-6.4. Two specific files showed `~5000ms` each on it.

Widening just the export type to `Record<string, { id: string; defaultMessage: string }>` (keeping the implementation literal) dropped both to <50ms.

**Downside, called out so nobody copy-pastes this without thinking:** we **lose compile-time literal-key checking** on every consumer of `sharedMessages`. A typo like `sharedMessages.fooBzr` no longer fails the build but returns `undefined` at runtime and we rely on tests / monitoring to catch it. The trade-off was acceptable to us because the literal-key inference was costing more wall-time than it was saving us bugs (we have other guards in place), but the cost is real.

For codebases that don't want to give that up wholesale, the same trick can be applied opt-in via a typed alias only at the hot call sites:

```ts
// cheap default export
export const sharedMessages: Record<
  string,
  { id: string; defaultMessage: string }
> = sharedMessagesInternal;

// per-file opt-in for callers that want the literal-key check back
type SharedMessageKey = keyof typeof sharedMessagesInternal;
const sharedMessages = sharedMessagesUntyped as Record<
  SharedMessageKey,
  { id: string; defaultMessage: string }
>;
```

That keeps the cheap export the default and lets individual files pay the inference cost only where it matters. Still under discussion on our side.

This isn't a styled-components problem on its own but on v6 the per-file wall-time tax made it visible.

### 3. Avoid `DeepPartial<HugeReduxState>` in test helpers

A test helper `setup(state?: DeepPartial<IState>, ...)` was the largest single-expression cost on the project (`checkDeferredNode` 4149ms). Every call site already cast its state to `Partial<IState>`. Replacing `DeepPartial` with `Partial` got it to 0ms — Types −1.3%, Instantiations −2.0% workspace-wide, from one parameter type change.

Same theme: any deep recursive type touched by many files becomes disproportionately expensive once the v6 base tax is in.

### 4. Restore v5's prop-forwarding behaviour for storybook tests

Storybook tests that scrape `console.error` started failing on every component that used native HTML elements with custom-cased props because v6 forwards everything to the DOM by default. We added a shared `shouldForwardProp` exported from our style system packe re-export module:

```ts
import isPropValid from '@emotion/is-prop-valid';

export function shouldForwardProp(propName: string, target: unknown): boolean {
  if (typeof target === 'string') {
    return !propName.startsWith('$') && isPropValid(propName);
  }
  return true;
}
```

…wired in once via `<StyleSheetManager shouldForwardProp={shouldForwardProp}>` at the top of each render tree (app, storybook, test-utils). Restores v5 behaviour, fixes the test failures, I think this is mentioned in the migration guide somewhere as well.

## TL;DR for fellow migrators (work-in-progress, but already directional)

The 6.4 type cost is real and not going away (per @quantizor's analysis above), but the practical fix-list is:

1. Move to `@typescript/native-preview` / tsgo — biggest single mechanical win.
2. Loosen `theme` typing in any styled-component helper modules to `any`.
3. Audit imports of huge literal-keyed objects; consider widening the export type, but be honest with yourself about losing literal-key checking on every consumer.
4. Audit usage of `DeepPartial<…>` over large state types; prefer `Partial<…>` where call sites already accept the looser shape.
5. Wire a v5-equivalent `shouldForwardProp` via `<StyleSheetManager>` at render-tree roots.

We're still on the migration branch but the combination of the above has us within striking distance of v5 type-check times on tsgo. Happy to share the audit doc / specific traces if useful.

**Comment by quantizor** [comment]
6.5 made a big improvement in this area, please give it a test https://github.com/styled-components/styled-components/releases/tag/styled-components%406.5.0
